src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -1927,36 +1927,52 @@ public DynamicObject sortVeryShort(VirtualFrame frame, DynamicObject array, NotP
+        public Object sortIntArray(DynamicObject array, NotProvided block,
+                @Cached("of(array)") ArrayStrategy strategy,
+                @Cached("basicIntStrategy()") ArrayStrategy intArrayStrategy) {
             final int size = getSize(array);
-            int[] copy = ((int[]) getStore(array)).clone();
+            ArrayMirror oldMirror = strategy.newMirror(array);
+            ArrayMirror newMirror = intArrayStrategy.newArray(size);
+            oldMirror.copyTo(newMirror, 0, 0, size);
+            int[] copy = (int[]) newMirror.getArray();
             doSort(copy, size);
             return createArray(copy, size);
         }
 
-        @Specialization(guards = { "!isEmptyArray(array)", "!isSmall(array)", "isLongArray(array)" },
-                        assumptions = "getContext().getCoreMethods().fixnumCmpAssumption")
-        public Object sortLongArray(DynamicObject array, NotProvided block) {
+        @Specialization(guards = { "forLong(strategy)", "!isEmptyArray(array)", "!isSmall(array)",
+                "strategy.matches(array)" }, assumptions = "getContext().getCoreMethods().fixnumCmpAssumption")
+        public Object sortLongArray(DynamicObject array, NotProvided block,
+                @Cached("of(array)") ArrayStrategy strategy,
+                @Cached("basicLongStrategy()") ArrayStrategy longArrayStrategy) {
             final int size = getSize(array);
-            long[] copy = ((long[]) getStore(array)).clone();
+            ArrayMirror oldMirror = strategy.newMirror(array);
+            ArrayMirror newMirror = longArrayStrategy.newArray(size);
+            oldMirror.copyTo(newMirror, 0, 0, size);
+            int[] copy = (int[]) newMirror.getArray();
             doSort(copy, size);
             return createArray(copy, size);
         }
 
-        @Specialization(guards = { "!isEmptyArray(array)", "!isSmall(array)", "isDoubleArray(array)" },
-                        assumptions = "getContext().getCoreMethods().floatCmpAssumption")
-        public Object sortDoubleArray(DynamicObject array, NotProvided block) {
+        @Specialization(guards = { "forDouble(strategy)", "!isEmptyArray(array)", "!isSmall(array)",
+                "strategy.matches(array)" }, assumptions = "getContext().getCoreMethods().fixnumCmpAssumption")
+        public Object sortDoubleArray(DynamicObject array, NotProvided block,
+                @Cached("of(array)") ArrayStrategy strategy,
+                @Cached("basicDoubleStrategy()") ArrayStrategy doubleArrayStrategy) {
             final int size = getSize(array);
-            double[] copy = ((double[]) getStore(array)).clone();
+            ArrayMirror oldMirror = strategy.newMirror(array);
+            ArrayMirror newMirror = doubleArrayStrategy.newArray(size);
+            oldMirror.copyTo(newMirror, 0, 0, size);
+            int[] copy = (int[]) newMirror.getArray();
             doSort(copy, size);
             return createArray(copy, size);
         }
```

Comment:
To avoid so much duplication around here, I think we should have `ArrayStrategy#isPrimitive()` and move the `Arrays.sort` boundary calls into the mirrors.
Then we could have only one specialization for primitives and one for non-primitives.